relays/ethereum/src/ethereum_exchange.rs

```diff
@@ -0,0 +1,211 @@
+use std::time::Duration;
+
+/// Interval at which we ask Ethereum node for updates.
+const ETHEREUM_TICK_INTERVAL: Duration = Duration::from_secs(10);
+/// Interval at which we ask Substrate node for updates.
+const SUBSTRATE_TICK_INTERVAL: Duration = Duration::from_secs(5);
+
+/// PoA exchange transaction relay params.
+#[derive(Debug, Default)]
+pub struct EthereumExchangeParams {
+	/// Ethereum connection params.
+	pub eth: EthereumConnectionParams,
+	/// Hash of the Ethereum transaction to relay.
+	pub eth_tx_hash: EthereumTransactionHash,
+	/// Substrate connection params.
+	pub sub: SubstrateConnectionParams,
+	/// Substrate connection params.
+	pub sub_sign: SubstrateSigningParams,
+}
+
+/// Ethereum to Substrate exchange pipeline.
+struct EthereumToSubstrateExchange;
+
+impl TransactionProofPipeline for EthereumToSubstrateExchange {
+	const SOURCE_NAME: &'static str = "Ethereum";
+	const TARGET_NAME: &'static str = "Substrate";
+
+	type BlockHash = H256;
+	type BlockNumber = u64;
+	type TransactionHash = EthereumTransactionHash;
+	type Transaction = EthereumTransaction;
+	type TransactionProof = EthereumTransactionInclusionProof;
+}
+
+/// Ethereum node as transactions proof source.
+struct EthereumTransactionsSource {
+	client: EthereumRpcClient,
+}
+
+#[async_trait]
+impl SourceClient<EthereumToSubstrateExchange> for EthereumTransactionsSource {
+	type Error = RpcError;
+
+	async fn tick(&self) {
+		async_std::task::sleep(ETHEREUM_TICK_INTERVAL).await;
+	}
+
+	async fn transaction(
+		&self,
+		hash: &EthereumTransactionHash,
+	) -> Result<Option<(EthereumHeaderId, EthereumTransaction)>, Self::Error> {
+		let eth_tx = match self.client.transaction_by_hash(*hash).await? {
+			Some(eth_tx) => eth_tx,
+			None => return Ok(None),
+		};
+
+		// we need transaction to be mined => check if it is included in the block
+		let eth_header_id = match (eth_tx.block_number, eth_tx.block_hash) {
+			(Some(block_number), Some(block_hash)) => HeaderId(block_number.as_u64(), block_hash),
+			_ => return Ok(None),
+		};
+
+		Ok(Some((eth_header_id, eth_tx)))
+	}
+
+	async fn transaction_proof(
+		&self,
+		eth_header_id: &EthereumHeaderId,
+		eth_tx: EthereumTransaction,
+	) -> Result<EthereumTransactionInclusionProof, Self::Error> {
+		let eth_header = self.client.header_by_hash(eth_header_id.1).await?;
+		let eth_relay_tx_hash = eth_tx.hash;
+		let mut eth_relay_tx = Some(eth_tx);
+		let mut eth_relay_tx_index = None;
+		let mut transaction_proof = Vec::with_capacity(eth_header.transactions.len());
+		for (index, eth_tx_hash) in eth_header.transactions.into_iter().enumerate() {
+			if eth_tx_hash != eth_relay_tx_hash {
+				let eth_tx = self.client.transaction_by_hash(eth_tx_hash).await?;
+				transaction_proof.push(match eth_tx.and_then(|eth_tx| eth_tx.raw) {
+					Some(eth_raw_tx) => eth_raw_tx.0,
+					None => return Err(EthereumNodeError::MissingTransaction(eth_tx_hash).into()),
```

Comment:
This error would no longer be required if we fetch block with transactions in one go.